lnwallet/wallet.go

```diff
@@ -992,44 +996,78 @@ func (l *LightningWallet) handleFundingCounterPartySigs(msg *addCounterPartySigs
 				txin.Witness[len(txin.Witness)-1],
 			)
 			if err != nil {
-				msg.err <- fmt.Errorf("cannot create script: "+
-					"%v", err)
-				msg.completeChan <- nil
-				return
+				return fmt.Errorf("cannot create script: %v", err)
 			}
 
 			output, err := l.Cfg.ChainIO.GetUtxo(
 				&txin.PreviousOutPoint,
 				pkScript, 0, l.quit,
 			)
 			if output == nil {
-				msg.err <- fmt.Errorf("input to funding tx "+
-					"does not exist: %v", err)
-				msg.completeChan <- nil
-				return
+				return fmt.Errorf("input to funding tx does "+
+					"not exist: %v", err)
 			}
 
 			// Ensure that the witness+sigScript combo is valid.
-			vm, err := txscript.NewEngine(output.PkScript,
-				fundingTx, i, txscript.StandardVerifyFlags, nil,
-				fundingHashCache, output.Value)
+			vm, err := txscript.NewEngine(
+				output.PkScript, fundingTx, i,
+				txscript.StandardVerifyFlags, nil,
+				fundingHashCache, output.Value,
+			)
 			if err != nil {
-				msg.err <- fmt.Errorf("cannot create script "+
+				return fmt.Errorf("cannot create script "+
 					"engine: %s", err)
-				msg.completeChan <- nil
-				return
 			}
 			if err = vm.Execute(); err != nil {
-				msg.err <- fmt.Errorf("cannot validate "+
+				return fmt.Errorf("cannot validate "+
 					"transaction: %s", err)
-				msg.completeChan <- nil
-				return
 			}
 
 			sigIndex++
 		}
 	}
 
+	return nil
+}
+
+// handleFundingCounterPartySigs is the final step in the channel reservation
+// workflow. During this step, we validate *all* the received signatures for
+// inputs to the funding transaction. If any of these are invalid, we bail,
+// and forcibly cancel this funding request. Additionally, we ensure that the
+// signature we received from the counterparty for our version of the commitment
+// transaction allows us to spend from the funding output with the addition of
+// our signature.
+func (l *LightningWallet) handleFundingCounterPartySigs(msg *addCounterPartySigsMsg) {
+	l.limboMtx.RLock()
+	res, ok := l.fundingLimbo[msg.pendingFundingID]
+	l.limboMtx.RUnlock()
+	if !ok {
+		msg.err <- fmt.Errorf("attempted to update non-existent funding state")
+		return
+	}
+
+	// Grab the mutex on the ChannelReservation to ensure thread-safety
+	res.Lock()
+	defer res.Unlock()
+
+	// Now we can complete the funding transaction by adding their
+	// signatures to their inputs.
+	res.theirFundingInputScripts = msg.theirFundingInputScripts
+	inputScripts := msg.theirFundingInputScripts
+
+	// Only if we have the final funding transaction do we need to verify
+	// the final set of inputs. Otherwise, it may be the case that the
+	// channel was funded via an external wallet.
+	fundingTx := res.fundingTx
```

Comment:
can move this inside the `if` since it will be `nil` otherwise

Reply:
Can move the below `res.partialState.FundingTxn = fundingTx` inside the if as well.

Reply:
Kept as is since we want to ensure external consistency (the full funding transaction is only available on the reservation struct after all other checks have passed). Only in the second instance is the variable actually accessed as is.